### PR TITLE
Add Dynamic Memory Sparsification (DMS) training and inference implementation

### PR DESCRIPTION
**Pull request by kstaniszewsknv**
## What does this PR do?

**Type of change:**  new feature

**Overview:** Training and inference code for Dynamic Memory Sparsification (DMS) - method from NeurIPS 2025 paper [Inference-Time Hyper-Scaling with KV Cache Compression](https://neurips.cc/virtual/2025/loc/san-diego/poster/119605)

## Usage
Detailed in `experimental/dms/README.md` and `experimental/dms/ARCHITECTURE.md`

## Testing
DMS tests in `experimental/dms/tests` covering:
* prefill
* generation
* gradient propagation
* chunked prefill

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: Yes
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: No, DMS is currently experimental feature with description in `experimental/dms`

## Additional Information
A minimal, optimized implementation of the DMS algorithm for KV-cache compression, as described in:

> **Inference-Time Hyper-Scaling with KV Cache Compression**  
> Adrian Łańcucki, Konrad Staniszewski, Piotr Nawrot, Edoardo M. Ponti  
> Paper: [https://arxiv.org/abs/2506.05345](https://arxiv.org/abs/2506.05345)  
> NeurIPS: [https://neurips.cc/virtual/2025/loc/san-diego/poster/119605](https://neurips.cc/virtual/2025/loc/san-diego/poster/119605)  

Inference-time scaling trades efficiency for improved reasoning by generating longer sequences. In Transformer LLMs, generation cost is often bottlenecked by the size of the key-value (KV) cache. DMS addresses this by learning a KV cache eviction policy that compresses the cache while preserving accuracy.

## How it works

DMS learns a per-head eviction policy that determines which KV cache entries to keep during generation. Rather than immediately discarding tokens, DMS delays eviction decisions, implicitly merging representations and preserving critical information. During training, the compression ratio is gradually increased from 1× to a target value (e.g., 8×), using knowledge distillation to match the outputs of an uncompressed teacher model.

## Summary by CodeRabbit

* **New Features**
  * Introduces Dynamic Memory Sparsification (DMS), an algorithm for efficient LLM inference and training with adaptive attention gating.
  * Adds DMS-enabled Qwen3 models with memory-efficient KV cache management and paged block-based storage.
  * Includes student-teacher distillation training infrastructure with noise scheduling and compression ratio control.
  * Provides configuration system and training/evaluation scripts for DMS adaptation.

* **Documentation**
  * Added architecture guide, README, and example inference notebook.

* **Tests**
  * Added comprehensive test suite for chunked prefill, cache management, and prefill/inference validation.

### CODE REVIEW

File: experimental/dms/dms/core.py [hunk 1]
```diff
@@ -0,0 +1,405 @@
...
+    hidden_states: torch.Tensor,
+    attention_mask: torch.Tensor | None,
+    position_ids: torch.Tensor | None,
+    past_key_values: Any,
+    use_cache: bool,
+    cache_position: torch.Tensor | None,
+    position_embeddings: torch.Tensor | None,
+    dms_manual_inference_mode: bool,
+    dms_chunked_prefill: int | None,
+    **kwargs: Any,
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/dms/core.py]: Fixed in commit [2d580d5](https://github.com/NVIDIA/Model-Optimizer/pull/877/commits/2d580d52b8538208ce09980f0fd93afe6bc80a0d)

File: experimental/dms/dms/training/engine.py [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+        student_model: PreTrainedModel,
+        teacher_model: PreTrainedModel,
+        trainer_state: DMSTrainerState,
+        dms_schedule: Callable[[int], float],
+        forward_fn: Callable,
+        student_is_teacher: bool,
+        tokenizer: PreTrainedTokenizer,
+        process_vocab_using_chunk: int,
+        forward_fn_kwargs_student: dict[str, Any] = {},
+        forward_fn_kwargs_teacher: dict[str, Any] = {},
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/dms/training/engine.py]: Alleviated in [c49701f](https://github.com/NVIDIA/Model-Optimizer/pull/877/commits/c49701f4c68e6d521e450c06c4e836fe3a298135)

**Comment by kevalmorabia97** [comment]
@kstaniszewsknv can you please confirm if all the code is NV written, and does not copy code from any 3rd party GitHub repositories?

File: experimental/dms/dms/attention.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        within_sliding_window = q_idx - k_idx <= window_size
+
+        causal = causal.to(score.dtype)
+        within_sliding_window = within_sliding_window.to(score.dtype)
+
+        modified_score = within_sliding_window * score + (1 - within_sliding_window) * (
+            dms_mask_values[b, k_idx] + score
+        )
+
+        return (1 - causal) * MASK_VALUE + causal * modified_score
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/dms/attention.py:157]: This is a desired behavior, as due to compatibility with already published checkpoints `DMSCache` uses `window_size + 1` in inference - see [this line](https://github.com/NVIDIA/Model-Optimizer/blob/c49701f4c68e6d521e450c06c4e836fe3a298135/experimental/dms/models/qwen3/modeling_qwen3_dms.py#L383).

File: experimental/dms/dms/training/data.py [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+        ds_sample = self.datasets[ds_idx][ds_sample_idx]
+        self.dataset_iterators[ds_idx] += 1
+
+        input_ids = np.array(ds_sample["encoded_prompt"], dtype=np.int64)
+
+        ds_sample_augmented = {
+            "input_ids": input_ids,
+            "attention_mask": np.ones_like(input_ids, dtype=bool),
+        }
+        return ds_sample_augmented
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/dms/training/data.py:459]: Not a problem since DataBlend is safely changed to the dataset [here](https://github.com/NVIDIA/Model-Optimizer/blob/c49701f4c68e6d521e450c06c4e836fe3a298135/experimental/dms/models/qwen3/train.py#L141). Code is still in experimental state and interface can change.

File: experimental/dms/models/qwen3/configuration_qwen3_dms.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        )
+        assert self.dms_window_size > self.dms_paged_attention_block_size, (
+            f"dms_window_size: {self.dms_window_size} "
+            f"is not greater than dms_paged_attention_block_size: {self.dms_paged_attention_block_size}"
+        )
+        assert self.dms_alpha_per in ["head", "layer"], (
+            f"dms_alpha_per: {self.dms_alpha_per} is not supported"
+        )
+        if dms_compile_limit is not None:
+            setup_compile_limit_for_dms(compile_limit=dms_compile_limit)
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/models/qwen3/configuration_qwen3_dms.py:98]: This executes only on loading/creation of config for a DMS enabled model. A suitable notification is already displayed  [setup_compile_limit_for_dms](https://github.com/NVIDIA/Model-Optimizer/blob/c49701f4c68e6d521e450c06c4e836fe3a298135/experimental/dms/dms/core.py#L42).

File: experimental/dms/models/qwen3/extract.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    cli_args = parser.parse_args()
+
+    cfg = load_config(cli_args.config)
+    model_cfg = cfg["model"]
+    dms_cfg = cfg["dms"]
+    data_cfg = cfg["data"]
+
+    checkpoint_dir = Path(cli_args.checkpoint)
+    model_path = checkpoint_dir / "pytorch_model.bin"
+    save_path = cli_args.output or str(checkpoint_dir / "student_model")
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/models/qwen3/extract.py:64]: Expected, as training configs explicitly save in legacy format.

File: experimental/dms/models/qwen3/extract.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        model_args=model_args,
+        training_args=training_args,
+        dms_cfg=dms_cfg,
+        data_cfg=data_cfg,
+        tokenizer=tokenizer,
+        trainer_state=trainer_state,
+    )
+
+    logger.info("Loading checkpoint weights...")
+    combined_model.load_state_dict(torch.load(model_path, weights_only=True))
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/models/qwen3/extract.py:96]: Expected, as training configs explicitly save in legacy format.

File: experimental/dms/models/qwen3/train.py [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+            output_dir.glob("checkpoint-*"),
+            key=lambda p: int(p.name.split("-")[1]),
+        )
+        if checkpoints:
+            logger.info(f"Auto-detected latest checkpoint: {checkpoints[-1]}")
+            return str(checkpoints[-1])
+        else:
+            logger.info("No checkpoints found, starting fresh.")
+            return None
+    return resume
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/models/qwen3/train.py:97]: This is expected behavior.

File: experimental/dms/models/qwen3/train.py [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+    config["auto_map"] = AUTO_MAP_CONFIG
+    with open(config_path, "w") as f:
+        json.dump(config, f, indent=2)
+
+    # Copy model implementation files for trust_remote_code
+    model_dir = Path(__file__).parent
+    for src_name in ["configuration_qwen3_dms.py", "modeling_qwen3_dms.py"]:
+        shutil.copy(model_dir / src_name, Path(save_path) / src_name)
+
+    logger.info(f"Successfully saved student model to: {save_path}")
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/models/qwen3/train.py:253]: Expected behavior.

File: experimental/dms/scripts/evaluate.sh [hunk 10]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
[the remark this answers is not in the document]
Comment (kstaniszewsknv) [reply experimental/dms/scripts/evaluate.sh:3]: This year was added by pre-commit checks.

**Comment by kstaniszewsknv** [comment]
> @kstaniszewsknv can you please confirm if all the code is NV written, and does not copy code from any 3rd party GitHub repositories?

The only code that we partially share with HF Transformers library is in `experimental/dms/models/qwen3/configuration_qwen3_dms.py` and `experimental/dms/models/qwen3/modeling_qwen3_dms.py` (Apache 2.0). 
We have already released a longer version of those two files here [https://huggingface.co/nvidia/Qwen3-8B-DMS-8x/tree/main](https://huggingface.co/nvidia/Qwen3-8B-DMS-8x/tree/main).
Those two files contain appropriate license notes (consulted with legal team).

File: experimental/dms/models/qwen3/configuration_qwen3_dms.py
Comment (kevalmorabia97) [inline experimental/dms/models/qwen3/configuration_qwen3_dms.py:1]: The license is a bit messed up. Correct order is 

```
# Adapted from <hf_link_with_commit_hash>

# Original Alibaba License

# NVIDIA Apache-2 Copyright and License
```

Please explude this file for auto-license header in https://github.com/NVIDIA/Model-Optimizer/blob/main/.pre-commit-config.yaml#L73-L99 and then manually fix the ordering of the file. See this file for reference: https://github.com/NVIDIA/Model-Optimizer/blob/main/examples/llm_eval/lm_eval_hf.py

File: experimental/dms/models/qwen3/modeling_qwen3_dms.py
Comment (kevalmorabia97) [inline experimental/dms/models/qwen3/modeling_qwen3_dms.py:1]: same comment

File: experimental/dms/models/qwen3/configuration_qwen3_dms.py
Comment (kstaniszewsknv) [reply experimental/dms/models/qwen3/configuration_qwen3_dms.py:1]: Thanks! Alleviated in [30c3c48](https://github.com/NVIDIA/Model-Optimizer/pull/877/commits/30c3c4850352039a1313293702270a2dd39fc763)

File: experimental/dms/models/qwen3/modeling_qwen3_dms.py
Comment (kstaniszewsknv) [reply experimental/dms/models/qwen3/modeling_qwen3_dms.py:1]: Thanks! Alleviated in [30c3c48](https://github.com/NVIDIA/Model-Optimizer/pull/877/commits/30c3c4850352039a1313293702270a2dd39fc763)

**Review by kevalmorabia97 — APPROVED** [review]

**Comment by BitPhinix** [comment]
Are there any plans to integrate this with an actual inference engine, such as trtllm / vllm?
